openwisp_notifications/views.py

```diff
@@ -42,4 +54,93 @@ def test_func(self):
         return True
 
 
-notification_preference_page = NotificationPreferencePage.as_view()
+@method_decorator(csrf_exempt, name='dispatch')
+class UnsubscribeView(TemplateView):
+    template_name = 'openwisp_notifications/unsubscribe.html'
+
+    def dispatch(self, request, *args, **kwargs):
+        self.encoded_token = request.GET.get('token')
+        if not self.encoded_token:
+            if request.method == 'POST':
+                return JsonResponse(
+                    {'success': False, 'message': 'No token provided'}, status=400
+                )
+            return render(request, self.template_name, {'valid': False})
+
+        self.user, self.valid = self._validate_token(self.encoded_token)
+        if not self.valid:
+            if request.method == 'POST':
+                return JsonResponse(
+                    {'success': False, 'message': 'Invalid or expired token'},
+                    status=400,
+                )
+            return render(request, self.template_name, {'valid': False})
+
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        is_subscribed = self.get_user_preference(self.user) if self.valid else False
+        context.update(
+            {
+                'valid': self.valid,
+                'user': self.user,
+                'is_subscribed': is_subscribed,
+            }
+        )
+        return context
+
+    def post(self, request, *args, **kwargs):
+        try:
+            if request.content_type == 'application/json':
+                data = json.loads(request.body)
+                subscribe = data.get('subscribe', False)
```

Comment:
```suggestion
                subscribe = data.get('subscribe', False) is True
```

This will ensure that  we don't pass any malformed data to the database query, e.g.:

```
{ 'subscribe': 'malformed'}
```

Reply:
Understood, Updated.